### PR TITLE
Avoid leakage of the workfile_set

### PR DESCRIPTION
**Pull request by gfphoenix78**
workfile_set is allocated from the shared memory and released back
to the shared memory in WorkFileDeleted(), mainly called in
AtProcExit_WorkFile(). The problem is only registered workfile_set
could be released. Before calling RegisterFileWithSet(), the allocated
workfile_set and its perquery is dangling and they could be leaked
if there is any ERROR/FATAL.
To avoid leakage, we track the allocated workfile_set in a local
array. In AtProcExit_WorkFile(), any dangling workfile_set is safely
released.

Reported-by: dh-cloud in PR #10186

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
Be adding test cases.

**Comment by dh-cloud** [comment]
The localEntries already keep track of the sets by the file-descriptor register to it, just a little problem to the no-file-register set  which i tried to fix it by registe a never used file-descriptor. 

However you designed another array just like the localEntries but only keep track of the sets, dont you thinks its a waste?
Maybe you'd better rewrite the code of keep-track-sets.

**Comment by gfphoenix78** [comment]
> The localEntries already keep track of the sets by the file-descriptor register to it, just a little problem to the no-file-register set which i tried to fix it by registe a never used file-descriptor.
> 
> However you designed another array just like the localEntries but only keep track of the sets, dont you thinks its a waste?
> Maybe you'd better rewrite the code of keep-track-sets.

`localEntries` is mainly used to quickly map the file descriptor to the workfile_set and the size value. The workfile_set in `localEntries` or WorkFileLocalEntry could be shared among different `localEntries`, which means the size of `localWorkSets` is less than that of `localEntries`, maybe much smaller.
The array here to keep track of workfile_set is similar to the ResourceOwner, but ResourceOwner manages query-related resources.
I thought about your suggestion before this PR, and didn't find a better way.

**Comment by dh-cloud** [comment]
I still think that register a file to the no-file-register work_set is much better.  I've read and test the  Virtual File Descriptor Alloc and comfirmed that 0 is never used, of cause there are also other ways to ensure that we register a file which not used before, such as in the `ensureLocalEntriesSize` we can let the `sizeLocalEntries` always bigger than the file then we use it......
Never mind, thank you for your review and i've fixed the `1` problem on my own project, done with this issue.
At Last, open your mind and there is always better way. This PR is too heavy that we have to maintenace another track-of-sets code.

File: src/test/isolation2/input/workfile_mgr_test.source [hunk 1]
```diff
@@ -0,0 +1,43 @@
+CREATE OR REPLACE FUNCTION gp_workfile_mgr_test_on_master(testname text)
```
Comment (xiong-gang) [inline src/test/isolation2/input/workfile_mgr_test.source:1]: please add some comments to explain how the test works, what are we expecting and what is the old behavior, etc.

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 2]
```diff
@@ -103,6 +106,62 @@ static bool proc_exit_hook_registered = false;
 Datum gp_workfile_mgr_cache_entries(PG_FUNCTION_ARGS);
 Datum gp_workfile_mgr_used_diskspace(PG_FUNCTION_ARGS);
 
+static void recycleWorkFileSet(workfile_set *work_set);
+static void ensureLocalWorkFileSet(void);
+static void forgetWorkFileSet(workfile_set *work_set);
+
+/* cleanup dangling workfile set in error recovery */
+void
+workfileCleanup(void)
```
Comment (xiong-gang) [inline src/backend/utils/workfile_manager/workfile_mgr.c:115]: dead code?

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 3]
```diff
@@ -103,6 +106,62 @@ static bool proc_exit_hook_registered = false;
...
+		localWorkSets = palloc(new_cap);
+	MemoryContextSwitchTo(old);
+	capLocalWorkSets = cap;
+}
+
+static void
+forgetWorkFileSet(workfile_set *work_set)
+{
+	int i;
+	for (i = sizeLocalWorkSets - 1; i >= 0; i--)
```
Comment (xiong-gang) [inline src/backend/utils/workfile_manager/workfile_mgr.c:149]: just curious, why do you prefer to iterate from the upper to the lower? there are usually several empty entries at the upper side, wouldn't that be a waste of cpu cycles?

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 4]
```diff
@@ -181,6 +240,26 @@ AtProcExit_WorkFile(int code, Datum arg)
...
+	LWLockAcquire(WorkFileManagerLock, LW_EXCLUSIVE);
+	for (i = sizeLocalWorkSets - 1; i >= 0; i--)
+	{
+		workfile_set *work_set = localWorkSets[i];
+		WorkFileUsagePerQuery *perquery;
+		ereport(WARNING,
+				(errmsg("leak workfile_set %p", work_set)));
+		perquery = work_set->perquery;
+
+		Assert(work_set->active);
```
Comment (xiong-gang) [inline src/backend/utils/workfile_manager/workfile_mgr.c:254]: in the case that there's no error out after work_set is created, the work_set is properly registered and delete in `WorkFileDeleted`, do we still need to call `recycleWorkFileSet` on it?

File: src/test/isolation2/input/workfile_mgr_test.source (hunk 1 above)
Comment (gfphoenix78) [reply src/test/isolation2/input/workfile_mgr_test.source:1]: The test file is moved from regress. Only the last 6 lines are added by this test.
The test is a wrapper for the tests in `workfile_mgr_test.c`.

The last 6 lines test that the leaked workfile_set is back to the shared memory.

File: src/backend/utils/workfile_manager/workfile_mgr.c (hunk 4 above)
Comment (gfphoenix78) [reply src/backend/utils/workfile_manager/workfile_mgr.c:238]: Yes, it gives back the leaked workfile_set to the shared memory. Fo example,
```c
work_set = workfile_mgr_create_set(NULL, NULL);
// before calling RegisterFileWithSet()
elog(ERROR, ...); // oops, the work_set is leaked. read new commands
work_set2 = workfile_mgr_create_set(NULL, NULL);
...
```

The for loop to run `WorkFileDeleted()` only gives the registered workfile_set back to the shared memory.

**Comment by gfphoenix78** [comment]
Close this PR in favor of https://github.com/greenplum-db/gpdb/pull/10719

**Pull request closed without merging**